### PR TITLE
Adding create task functionality

### PR DESCRIPTION
**Pull request by Vatsal09**
[Feature]: Added tasks (specifically create task) to the API Gateway

### Description

Added the ability to expose internal task service to the public through the API Gateway

In this PR:
- [x] Added configs for task service to the gateway
- [x] added create task handlers
- [x] added documents used by the TaskResponse
- [x] added clients (public-facing) task document
- [x] tests for handlers and documents for create task

### Reasoning

_Why did you make this change?_

To enable mobile app to create tasks

### Dependencies

_Any dependent PRs/conditions we need to be aware of?_

No

### Testing

_How do we know this works?_

- [x]  unit tests 
- [x]  smoke tests 
- [x] system testing

### CODE REVIEW

File: src/clients/task.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    PROTOCOL = "http://"
+    TASK_ENDPOINT = "/api/v1/tasks"
+
+    @annotate_arg("hostport", with_annotation="TASK_SERVICE_HOSTPORT")
+    def __init__(self, hostport, http_conn_pool=None):
+        self.hostport = hostport
+        self.http_conn_pool = Session() if not http_conn_pool else http_conn_pool
+
+    def create(self, title, user_id, **kwargs):
+        print(user_id)
```
Comment (joshpaulchan) [inline src/clients/task.py]: Don't forget to remove this!

File: src/clients/task.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        post_payload = {"title": title, "userId": user_id}
+        for kw in kwargs:
```
Comment (joshpaulchan) [inline src/clients/task.py]: suggestion:

```
for optional_arg in ["due", "deleted", "completed"]:
    if optional_arg in kwargs:
         post_payload[optional_arg] = kwargs.get(optional_arg)
```

File: tests/tasks/test_documents.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+from src.tasks.documents import Task, TaskResponse
+
+
+def test_task_class_construction():
+    users = {1,2,3}
+    title = "Contact John Doe"
+    task = Task(title, users)
+
+    assert task.users == users
+    assert len(task.users) == 3
```
Comment (joshpaulchan) [inline tests/tasks/test_documents.py]: yeah this is redundant, `assert task.users == users` does everything you need

File: tests/tasks/test_documents.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    assert task.title == title
+
+def test_task_from_model_method():
+    task = Mock(**{"title": "some task title", "users": {24, 8}})
+    task_doc = Task.from_model(task)
+
+    assert task_doc.title == "some task title"
+    assert task_doc.users == {8, 24}
+    assert len(task_doc.users) == 2
```
Comment (joshpaulchan) [inline tests/tasks/test_documents.py]: same here

File: src/tasks/handlers.py [hunk 5]
```diff
@@ -1,12 +1,31 @@
...
 
     def __call__(self, request):
-        # request.user.id, request.user.email... see clients.auth.User for full list of User details
-        return jsonify([])
+
+        client_post_request = request.json
+        if not client_post_request:
+            return empty(status=HTTPStatus.BAD_REQUEST)
+
+        user = request.user.get("user")
```
Comment (joshpaulchan) [inline src/tasks/handlers.py]: when we merge master will probably have a few merge conflicts but you'll be able to do `user = request.user` (or just not do it at all)

File: tests/tasks/test_documents.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
+from unittest.mock import Mock
+
+import pytest
+
+from src.tasks.documents import Task, TaskResponse
+
+
+def test_task_class_construction():
```
Comment (joshpaulchan) [inline tests/tasks/test_documents.py]: looking good!

File: src/clients/task.py (hunk 1 above)
Comment (Vatsal09) [reply src/clients/task.py]: Yup, removed it! LOL

File: src/clients/task.py (hunk 2 above)
Comment (Vatsal09) [reply src/clients/task.py]: refactored the code a little here. you can recheck

File: tests/tasks/test_documents.py (hunk 4 above)
Comment (Vatsal09) [reply tests/tasks/test_documents.py]: ACK

File: tests/tasks/test_documents.py (hunk 3 above)
Comment (Vatsal09) [reply tests/tasks/test_documents.py]: ACK

File: src/tasks/handlers.py (hunk 5 above)
Comment (Vatsal09) [reply src/tasks/handlers.py]: I pulled your changes. So hopefully not many merge conflicts

**Comment by Vatsal09** [comment]
All code, tests, and PR changes are in. You can review it once again. @joshpaulchan

File: src/clients/task.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    @annotate_arg("hostport", with_annotation="TASK_SERVICE_HOSTPORT")
+    def __init__(self, hostport, http_conn_pool=None):
+        self.hostport = hostport
+        self.http_conn_pool = Session() if not http_conn_pool else http_conn_pool
+
+    def create(self, title, user_id, **kwargs):
+        
+        post_payload = {"title": title, "userId": user_id}
+        possible_task_args = ["due", "deleted", "completed"]
```
Comment (joshpaulchan) [inline src/clients/task.py]: NitPick: Maybe a better name could be `optional_task_args`? More importantly I think you probably want to make this a constant (either in the class like `PROTOCOL`, or in the file itself)

File: src/clients/task.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        add_args = (arg for arg in possible_task_args if arg in kwargs)
+        
+        for arg in add_args:
+            post_payload[arg] = kwargs.get(arg)
+
+        task_create_response = self.http_conn_pool.post(
+            self._build_url(self.TASK_ENDPOINT),
+            json=post_payload,
+        )
+        if task_create_response.status_code in range(400,500):
```
Comment (joshpaulchan) [inline src/clients/task.py]: I was wondering this before but I guess it's worth just double checking: How does this work? Does it have to expand this range and iterate through to see if we're in the range, or does it do something smarter more performant? (like do `400 < status code < 500` or something)

File: src/tasks/handlers.py [hunk 9]
```diff
@@ -1,12 +1,31 @@
...
 
     def __call__(self, request):
-        # request.user.id, request.user.email... see clients.auth.User for full list of User details
-        return jsonify([])
+
+        client_post_request = request.json
+        if not client_post_request:
+            return empty(status=HTTPStatus.BAD_REQUEST)
+
+        user = request.user
```
Comment (joshpaulchan) [inline src/tasks/handlers.py]: I forgot to mention, but the way things are implemented, `request.user` is only there if the user passes the token and it checks out - it's possible to have an unauthenticated request come through. You'll probably want to guard against that as well and add a test to see your guard works

File: tests/tasks/test_handlers.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+@pytest.fixture
+def task_svc():
+    return create_autospec(TaskService)
+
+@pytest.fixture
+def task_create_handler(task_svc):
+    return TaskCreateHandler(task_svc)
+
+def test_login_handler_missing_data_should_fail(task_create_handler):
```
Comment (joshpaulchan) [inline tests/tasks/test_handlers.py]: Yeah, something similar to a mix of the these two cases. Valid data but no user should probably return a `401` I think?

File: src/clients/task.py (hunk 7 above)
Comment (Vatsal09) [reply src/clients/task.py]: ACK

File: src/clients/task.py (hunk 8 above)
Comment (Vatsal09) [reply src/clients/task.py]: Performance-wise, it is just about the same since it's not creating an actual list but I can change it

File: src/tasks/handlers.py (hunk 9 above)
Comment (Vatsal09) [reply src/tasks/handlers.py]: I have tried several ways of trying access without a proper authentication but I have not been able to hit the task handler. Can you let me know which exact scenario you are referring to so I can guard against that?

Comment (Vatsal09) [reply src/tasks/handlers.py]: You pass a user = None, right? When the tokens are not valid?

File: src/clients/task.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+    @annotate_arg("hostport", with_annotation="TASK_SERVICE_HOSTPORT")
+    def __init__(self, hostport, http_conn_pool=None):
+        self.hostport = hostport
+        self.http_conn_pool = Session() if not http_conn_pool else http_conn_pool
+
+    def create(self, title, user_id, **kwargs):
+        
+        post_payload = {"title": title, "userId": user_id}
+
```
Comment (joshpaulchan) [inline src/clients/task.py]: NP: I think we need to get your formatter set up so it auto formats stuff (like removing unnecessary whitespace) but this is fine

File: src/tasks/handlers.py (hunk 9 above)
Comment (joshpaulchan) [reply src/tasks/handlers.py]: yeah double checking the code: https://github.com/getcastle/api-gateway/blob/master/src/common/routes.py#L81 seems we already do what I was talking about so it's properly named as `authenticated` and does its job. Knowing that, it's up to you whether we should still include the test case for unauthenticated user - up to you if you want to keep it or not!
